[LangRef] Rewrite the elementwise atomic description and reject seq_cst elementwise atomics (#209931)
llvm/llvm-project: llvm/docs/LangRef.md

monotonic modification order with other operations that are not marked
`syncscope("singlethread")` or `syncscope("<target-scope>")`.

(elementwise-atomics)=

### Elementwise Atomic Operations

Certain atomic instructions, such as {ref}`atomicrmw <i_atomicrmw>`,
and {ref}`atomic load <i_load>`, may be marked `elementwise`. The access type
must then be a fixed vector type whose total bit width is a power of two greater
than or equal to eight, and whose element type is supported by the corresponding
scalar atomic instruction. The {ref}`ordering <ordering>` of an `elementwise`
instruction may not be `seq_cst`.

For the purposes of the memory model, the instruction behaves as if one thread
were spawned per element, each performing the corresponding scalar atomic

> operation on that element, and then joining after completion. For the purposes

Reviewer: I might be missing something here, though I have been wondering how, say, an elementwise atomicrmw with monotonic ordering (which does not synchronize) would interact with a preceding fence release under this analogy. If the parent thread issues the fence, but the per-element operations would execute in "other spawned threads", would the fence be sequenced-before them (which is intra-thread) still hold?

Reviewer 2: Oh, that is a good point. In the model usually a release fence needs to be program-order before the relaxed write to establish fence-based release-acquire...

Reviewer: I think the analogy makes sense intuitively, I’m just not completely sure it holds up well enough to formalize/clarify the concept (esp. given the recent efforts to refine LangRef semantics…).

Reviewer 2: Yeah so basically what we want is that the per-element operations are program-order-after the previous instruction, and program-order-before the next instruction, but are not themselves ordered by program-order.

I don't see a way to express this without explicitly referring to the relevant order. (I don't care strongly whether it is called program-order or sequenced-before.)

Reviewer: To be clear, I was not advocating for one term over the other in the previous suggestion. I just feel that if we use sequenced-before, it may be worth introducing/explaining it first.

Author: > Yeah so basically what we want is that the per-element operations are program-order-after the previous instruction, and program-order-before the next instruction, but are not themselves ordered by program-order.

I like this definition also, I think this is closer to what I originally had. What is the problem with referring to the program order? The LangRef defines:

> We define a happens-before partial order as the least partial order that

> Is a superset of single-thread program order, and

> When a synchronizes-with b, includes an edge from a to b. Synchronizes-with pairs are introduced by platform-specific techniques, like pthread locks, thread creation, thread joining, etc., and by atomic instructions. (See also [Atomic Memory Ordering Constraints](https://llvm.org/docs/LangRef.html#ordering)).

So the LangRef already mentions "single-thread program order", I think we can define it exactly how Ralf just described?

Author: Ok I pushed another iteration. What do we think?

monotonic modification order with other operations that are not marked
`syncscope("singlethread")` or `syncscope("<target-scope>")`.

(elementwise-atomics)=

### Elementwise Atomic Operations

Certain atomic instructions, such as {ref}`atomicrmw <i_atomicrmw>`,
and {ref}`atomic load <i_load>`, may be marked `elementwise`. The access type
must then be a fixed vector type whose total bit width is a power of two greater
than or equal to eight, and whose element type is supported by the corresponding
scalar atomic instruction.

> For the purposes of the happens-before relation, the instruction behaves as if

Reviewer 2: It's not just hb, right? It's the same for rf (reads-from) and po (program-order, also sometimes called sequenced-before).
```suggestion
For the purposes of the memory model orders (happens-before, sequenced-before, etc), the instruction behaves as if
```

Reviewer: Considering that happens-before does encompass program order, and that the happens-before relation is defined a few lines earlier, I'd suggest not introducing C++ terms (sequenced-before) that LangRef doesn't otherwise use. If the shorter form doesn't suffice:

```suggestion
For the purposes of the memory model, the instruction behaves as if
```
Perhaps an explicit refinement could go along the following lines:
``` 
The scalar operations of a single elementwise instruction are not related to
one another by program order; consequently, the instruction does not, by
itself, establish any happens-before relationship among them.
```

Reviewer 2: Happens-before is fully defined in terms of other relations, so if this paragraph here was intending to change the definition of happens-before that'd be a pretty fundamental change. I also think C++ terms should be introduced somewhere since that's the terms the C++ model is defined in, and it's not like we have another more credible model we could use instead.

But I think I see your point. I am also fine with just "For the purposes of the memory model".

Author: Fixed, I'm good with "For the purposes of the memory model".

one thread were spawned per element, each performing the corresponding scalar
atomic operation on that element, and then joining after completion. For the
purposes of `syncscope`, each scalar operation is still considered to execute
in the same thread as the original instruction. Synchronizing with one of the
scalar operations does not, by itself, establish a happens-before relationship
with another scalar operation from the same `elementwise` instruction.

The instruction's `syncscope` and {ref}`ordering <ordering>` apply

> independently to each scalar operation and the `ordering` may not be `seq_cst`.

Reviewer 2: What does the "apply independently" here mean that's not already said in the previous paragraph?

Author: You're right, I don't think this part is meaningful.

Author: Removed this part.

monotonic modification order with other operations that are not marked
`syncscope("singlethread")` or `syncscope("<target-scope>")`.

(elementwise-atomics)=

### Elementwise Atomic Operations

Certain atomic instructions, such as {ref}`atomicrmw <i_atomicrmw>`,
and {ref}`atomic load <i_load>`, may be marked `elementwise`. The access type
must then be a fixed vector type whose total bit width is a power of two greater
than or equal to eight, and whose element type is supported by the corresponding

> scalar atomic instruction. The {ref}`ordering <ordering>` of an `elementwise`

Reviewer 3: This second part kinda overlaps with the first part, right? The first part only needs to say that the vector size is a power of two?

Author: Agreed, if the element type is supported by the scalar atomic instruction, then each element type must at least be `i8` and this implies the total bit width is at least `i8`.

instruction may not be `seq_cst`.

An `elementwise` atomic instruction behaves as if it were expanded into one
scalar version of that instruction for each vector element. Each resulting
scalar operation has the same {ref}`ordering <ordering>` and `syncscope` as the
original instruction. Each scalar operation occupies the original instruction's
position in program order relative to other operations, but the scalar

> operations are not ordered by program order with respect to one another.

Reviewer 3: ```suggestion
operations are not related in program order with respect to one another.
```

> Synchronizing with one scalar operation does not, by itself, establish a

Reviewer 2: ```suggestion
Synchronizing with one scalar operation therefore does not, by itself, establish a
```

floating-point, or vector type whose bit width is a power of two greater than or
equal to eight.

If the `load` is marked `elementwise`, the instruction has
{ref}`elementwise atomic semantics <elementwise-atomics>`. The loaded type must

> be a fixed vector type whose total bit width is a power of two greater than or

Reviewer 3: Same question about just the vector size being power of two.

"{ref}`syncscope <syncscope>`" argument.

If the `elementwise` modifier is present, the instruction has per-element vector
atomic semantics. It behaves as if it were expanded into one scalar `atomicrmw`
per element, that are not ordered with respect to each other. In other words, a
consistent ordering does not exist between the individual scalar operations of
the same `elementwise` instruction. Synchronizing with one of the scalar

> operations does not, by itself, establish a happens-before relationship with another scalar operation from the same `elementwise` instruction.

Reviewer 2: Is this equivalent to "spawn one thread per lane, each doing the access for one element, then join them all"?

Author: I think this sounds right? As long as we do not allow `seq_cst`, which under your definition would consistently order the accesses.

Reviewer 2: IMO that would be good to mention (or even to use as the definition).

Author: On second thought I'm not sure if I like the thread analogy because it has implications on `syncscope` like `syncscope("singlethread")`. I think we'd want to be clear this is all happening on the same thread.

I'm open to other wording recommendations though, but I think what I wrote is relatively clear (at least to me).

Author: One thing I'd change is `consistent ordering does not exist` to `consistent ordering is not required to exist` because stronger program models could use this.

Reviewer 2: >  it has implications on syncscope like syncscope("singlethread")

That's a good point, but that can be mentioned in the docs.

I find the current docs still kind of hard to interpret. The memory model has half a dozen orders, so saying "a consistent ordering does not exist" makes it unclear which of them you refer to an how exactly each one of them is affected.

Author: What do you think about the wording now?
